// tsparticles-editor/dist/Sections/Options/Particles/Opacity/OpacityOptionsEditor.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { EditorType } from "object-gui";
import { EditorBase } from "../../../../EditorBase";
export class OpacityOptionsEditor extends EditorBase {
    constructor(particles) {
        super(particles);
    }
    addToGroup(parent, options) {
        this.group = parent.addGroup("opacity", "Opacity", true, options);
        this.options = this.group.data;
        this.addAnimation();
        this.addRandom();
        this.addProperties();
    }
    addAnimation() {
        const particles = this.particles;
        const group = this.group.addGroup("animation", "Animation");
        group
            .addProperty("destroy", "Destroy", EditorType.select)
            .change(() => __awaiter(this, void 0, void 0, function* () {
            yield particles.refresh();
        }))
            .addItems([
            {
                value: "max",
            },
            {
                value: "min",
            },
            {
                value: "none",
            },
        ]);
        group.addProperty("enable", "Enable", EditorType.boolean).change(() => __awaiter(this, void 0, void 0, function* () {
            yield particles.refresh();
        }));
        group
            .addProperty("minimumValue", "Minimum Value", EditorType.number)
            .change(() => __awaiter(this, void 0, void 0, function* () {
            yield particles.refresh();
        }))
            .min(0)
            .max(0)
            .step(0.01);
        group
            .addProperty("speed", "Speed", EditorType.number)
            .change(() => __awaiter(this, void 0, void 0, function* () {
            yield particles.refresh();
        }))
            .step(0.01);
        group
            .addProperty("startValue", "Start Value", EditorType.select)
            .change(() => __awaiter(this, void 0, void 0, function* () {
            yield particles.refresh();
        }))
            .addItems([
            {
                value: "max",
            },
            {
                value: "min",
            },
            {
                value: "random",
            },
        ]);
        group.addProperty("sync", "Sync", EditorType.boolean).change(() => __awaiter(this, void 0, void 0, function* () {
            yield particles.refresh();
        }));
    }
    addRandom() {
        const particles = this.particles;
        const group = this.group.addGroup("random", "Random");
        group.addProperty("enable", "Enable", EditorType.boolean).change(() => __awaiter(this, void 0, void 0, function* () {
            yield particles.refresh();
        }));
        group.addProperty("minimumValue", "Minimum Value", EditorType.number).change(() => __awaiter(this, void 0, void 0, function* () {
            yield particles.refresh();
        }));
    }
    addProperties() {
        const particles = this.particles;
        this.group
            .addProperty("value", "Value", EditorType.number)
            .change(() => __awaiter(this, void 0, void 0, function* () {
            yield particles.refresh();
        }))
            .min(0)
            .max(1)
            .step(0.01);
    }
}
