// tsparticles-editor/dist/browser/Sections/Options/Particles/Twinkle/TwinkleOptionsEditor.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { EditorType } from "object-gui";
import { EditorBase } from "../../../../EditorBase";
export class TwinkleOptionsEditor extends EditorBase {
    constructor(particles) {
        super(particles);
    }
    addToGroup(parent, options) {
        this.group = parent.addGroup("twinkle", "Twinkle", true, options);
        this.options = this.group.data;
        this.addTwinkle();
    }
    addTwinkle() {
        this.addTwinkleValues(this.group.addGroup("lines", "Lines"));
        this.addTwinkleValues(this.group.addGroup("particles", "Particles"));
    }
    addTwinkleValues(group) {
        var _a;
        const particles = this.particles;
        const options = group.data;
        const color = typeof options.color === "string" ? options.color : (_a = options.color) === null || _a === void 0 ? void 0 : _a.value;
        group.addProperty("color", "Color", EditorType.color, color, false).change((value) => __awaiter(this, void 0, void 0, function* () {
            if (typeof value === "string") {
                if (typeof options.color === "string") {
                    options.color = value;
                }
                else {
                    options.color = {
                        value,
                    };
                }
            }
            yield particles.refresh();
        }));
        group.addProperty("enable", "Enable", EditorType.boolean).change(() => __awaiter(this, void 0, void 0, function* () {
            yield particles.refresh();
        }));
        group.addProperty("frequency", "Frequency", EditorType.number).change(() => __awaiter(this, void 0, void 0, function* () {
            yield particles.refresh();
        }));
        group
            .addProperty("opacity", "Opacity", EditorType.number)
            .change(() => __awaiter(this, void 0, void 0, function* () {
            yield particles.refresh();
        }))
            .step(0.01)
            .min(0)
            .max(1);
    }
}
